% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ATAC-seq.R
\name{find_motifs}
\alias{find_motifs}
\title{Find motifs}
\usage{
find_motifs(
  motif,
  step = 20,
  fimodir,
  outputdir1,
  outputdir,
  Motifdir,
  sequencedir
)
}
\arguments{
\item{motif}{motif file, you can choose our bulit-in motif database of
'mus musculus', 'homo sapiens', 'zebrafish' and 'chicken' by 'motif = Tranfac201803_Mm_MotifTFsF',
'motif = Tranfac201803_Hs_MotifTFsF', 'motif = Tranfac201803_Zf_MotifTFsF',
'motif = Tranfac201803_Ch_MotifTFsF' respectively, or you can upload your own
motif data base, but the formata use be the same as our built-in motif database.}

\item{step}{numeric, indicating the numbers of motif in each group. Because
there are so many motifs need to be calculated, so we divided every 20 motifs
into groups and then calculated each group using FIMO at the same time by nohup}

\item{fimodir}{character, indicating path of fimo software,
if you have added fimo to the environment variable, just set this argument as 'fimo'}

\item{outputdir1}{character, indicating the output path of shell script}

\item{outputdir}{character, indicating output path of fimo result}

\item{Motifdir}{character, indicating the path of meme motif file}

\item{sequencedir}{sequence file directory}
}
\value{
return scripts to run Fimo in command line
}
\description{
Generate shell commands to find motifs in the footprints through
fimo software.
}
\examples{
motif1 <- Tranfac201803_Mm_MotifTFsF
fimodir <- 'fimo'
outputdir <- 'D:/GIBH/IReNA2 R package/IReNA2/ATAC/outputdir/'
motifdir <- 'Mememotif/'
sequencedir <- 'merged_footprints.fasta'
#find_motifs(motif1,step=20,fimodir, outputdir, motifdir, sequencedir)
}
